### [1] Importing the libraries:

In [182]:
import pandas as pd
import numpy as np
import json
import shutup 
shutup.please()

##### [1.1]
##### What is importing:
Importing is the process of extracting the libraries from Python and using it in your source environment.
##### What are we importing:
For the basic summary calculation, and exploring the data, we have used numpy, pandas, json, and shutup
##### What is 'shutup':
Shutup is a library for removing all sorts of warnings and annoying messages. This is not inbuilt in Python and has to be installed seperately. I found this randomly when searching for removing warnings. 

Once can install shutup by running the following codeblock.

    !pip install shutup

We are only using shutup.please() function as it is most useful for removing the warnings.
Here is the link for the shutup library if you would like to read more:  [Shutup](https://github.com/polvoazul/shutup)
       


### [2] Reading the file

In [127]:
df = pd.read_json("transactions.txt", lines = True)

#### [2.1]

##### Description:
We are reading the file using a pandas function called "read_json". It reads files in a JSON format.

##### Input Parameters:
The input parameters for the function are:

The data which is stored in JSON format.

The json is in the lines-delimited format. Hence we also pass the lines variable:
        
    lines=True
Usually this is not required and the default is False.

##### Output:
The output for the read_json is a pandas DataFrame.

##### Suggestion:

Failed to read lines using:
    
    with open('data/nested_mix.json','r') as f:
            data = json.loads(f.read())
    
    df = pd.json_normalize(data, record_path =['students'])
    
Pandas.read_json is much more powerful and easier to read json files.
       

### [3] Displaying the DataFrame

In [128]:
pd.set_option('display.max_columns', None)

In [130]:
df.head()

accountNumber  customerId  creditLimit  availableMoney  \
0      737265056   737265056         5000          5000.0   
1      737265056   737265056         5000          5000.0   
2      737265056   737265056         5000          5000.0   
3      737265056   737265056         5000          5000.0   
4      830329091   830329091         5000          5000.0   

   transactionDateTime  transactionAmount         merchantName acqCountry  \
0  2016-08-13T14:27:32              98.55                 Uber         US   
1  2016-10-11T05:05:54              74.51          AMC #191138         US   
2  2016-11-08T09:18:39               7.47           Play Store         US   
3  2016-12-10T02:14:50               7.47           Play Store         US   
4  2016-03-24T21:04:46              71.18  Tim Hortons #947751         US   

  merchantCountryCode posEntryMode posConditionCode merchantCategoryCode  \
0                  US           02               01            rideshare   
1                  US           09               01        entertainment   
2                  US           09               01           mobileapps   
3                  US           09               01           mobileapps   
4                  US           02               01             fastfood   

  currentExpDate accountOpenDate dateOfLastAddressChange  cardCVV  enteredCVV  \
0        06/2023      2015-03-14              2015-03-14      414         414   
1        02/2024      2015-03-14              2015-03-14      486         486   
2        08/2025      2015-03-14              2015-03-14      486         486   
3        08/2025      2015-03-14              2015-03-14      486         486   
4        10/2029      2015-08-06              2015-08-06      885         885   

   cardLast4Digits transactionType echoBuffer  currentBalance merchantCity  \
0             1803        PURCHASE                        0.0                
1              767        PURCHASE                        0.0                
2              767        PURCHASE                        0.0                
3              767        PURCHASE                        0.0                
4             3143        PURCHASE                        0.0                

  merchantState merchantZip  cardPresent posOnPremises recurringAuthInd  \
0                                  False                                  
1                                   True                                  
2                                  False                                  
3                                  False                                  
4                                   True                                  

   expirationDateKeyInMatch  isFraud  
0                     False    False  
1                     False    False  
2                     False    False  
3                     False    False  
4                     False    False

#### [3.1]
##### Description:
We are using the head function to display the dataframe. An additonal pandas option is used to display all the columns as we are trying to figure out the parameters.

##### Input Parameters:
The dataframe and the pandas.set_option variable with the following arguments.

    'display.max_columns'
    None
##### Output:

The Pandas DataFrame object with all the columns visible. 

### [4] Data Exploration and summary statistics

#### [4.1] Finding  the number of records and fields in each record

In [131]:
df.shape

(786363, 29)

##### Description:
This is given by the shape of the data. The number of records is given by the columns of the shape and fields in each record is given by the rows of the shape.

##### Input Parameters:
Dataframe and the shape attribute.

##### Output:
Number of records in the dataframe are: 29

Number of fields in each record: 786363

#### [4.2] Addtional summary statistics for each column

Since the account number is a record which should not be having any other statistic than the unique value, I assume there will not be null, minimum or maximum.

In [134]:
print(len(df["accountNumber"].unique()))

5000


##### Description:
The above outputs the number of unique customers in the dataframe. 

##### Input Parameters:
Unique Function, The accountNumber record

##### Output:

Length of the unique number of records



Here there are 5000 unique customers in the dataframe.

#### [4.3]
**Assumption:** Comparing the customerId and AccountNumber as my assumption for the class is they both are the same.

In [158]:
count = 0
for i in range(len(df)):
    if(df.customerId[i]==df.accountNumber[i]):
        count+=1
print(count)

786363


##### Description:
Just as a check, the customerId is the same as the accountNumber, so it can be dropped in our further usage. 

##### Input Parameters:
The Length of the dataframe,

The customerID record,

The account number record

##### Output:

The number of record of equal CustomerID and AccountNumber.




Here every account number and the customerId is the same which is 786363. The customerId will be dropped in the further process

#### [4.4]

#### Unique Function:
We are using the unique function for the next records.

##### Description:
For different customers, it is observerd that there can be same CVVs but it because of the limited combinations, only 899 CVVs are found.

##### Input:
The record is passed onto the unique function.

##### Output:
The length of Unique CVV.

#### [4.4.1]

##### Let's check the unique number of CVV. 

**Assumption:** There will be <=900 unique CVVs. 9x10x10, assuming the first number doesn't start with a 0. There will be more enteredCVV because that column will be helpful for identifying fraud.

In [191]:
print(len(df.cardCVV.unique()))

899


In [192]:
print(len(df.enteredCVV.unique()))

976


In [150]:
len(df.cardLast4Digits.unique())

5245

#### Result:
This is an interesting result as it shows that there are more than 5000 cards. So it is possible that a few customers have different cards with different credit limits

In [193]:
len(df.transactionDateTime.unique())

776637

Most transaction times are unique except for the reversal and multiswipes

#### [4.5]

#### Describe and sum of is na Function:
We are using the Describe function and the summation of the na values for the records.



##### Description:
Outputtng the summary statsitics along with the NA values with for the following records
    
    availableMoney
    transactionAmount
    creditLimit

##### Input:
The record is passed onto the describe function. 

We use the record and apply the summation for isna() function.

##### Output:
The count, mean, standard deviation, minimum, maximum, percentile of Unique CVV.

The na values for the records

In [147]:
df["availableMoney"].describe()

count    786363.000000
mean       6250.725369
std        8880.783989
min       -1005.630000
25%        1077.420000
50%        3184.860000
75%        7500.000000
max       50000.000000
Name: availableMoney, dtype: float64

In [186]:
sum(df.availableMoney.isna())

0

In [139]:
df["transactionAmount"].describe()

count    786363.000000
mean        136.985791
std         147.725569
min           0.000000
25%          33.650000
50%          87.900000
75%         191.480000
max        2011.540000
Name: transactionAmount, dtype: float64

In [187]:
sum(df.transactionAmount.isna())

0

In [138]:
df["creditLimit"].describe()

count    786363.000000
mean      10759.464459
std       11636.174890
min         250.000000
25%        5000.000000
50%        7500.000000
75%       15000.000000
max       50000.000000
Name: creditLimit, dtype: float64

In [188]:
sum(df.creditLimit.isna())

0

All of the above values have the same NA values, which is zero.

Although, the Creditlimit is not sorted for every customer. Lets use the groupby function for a clear measure of the creditLimit for each customer.

In [144]:
creditLimit_df = df.groupby("accountNumber")["creditLimit"].mean().to_frame()

##### [4.5.1]

##### Description:
Using the groupby function for grouping according to the mean creditLimit for the number of transactions of a customer.

##### Input:
The records passed into the groupby function:
    
    accountNumber
    creditLimit

The mean function is used:
    
##### Output:
 A dataframe which is grouped according to the accountNumber

In [146]:
creditLimit_df.creditLimit.describe()

count     5000.000000
mean     10149.150000
std      10816.765564
min        250.000000
25%       5000.000000
50%       7500.000000
75%      15000.000000
max      50000.000000
Name: creditLimit, dtype: float64

This is a better representation for the creditLimit for the customer.

#### [4.6]

**Assumption:** The NA values in the a few records are hidden and should be used a different method to find.

#### Value Counts:
We are using the Value Counts function for identifying the null values in certain records.



##### Description:
Outputting the  NA values for the records
    

##### Input:
The record is passed into the value_counts function
##### Output:
The number of  na values for each record

In [152]:
df["transactionType"].value_counts()

PURCHASE                745193
REVERSAL                 20303
ADDRESS_VERIFICATION     20169
                           698
Name: transactionType, dtype: int64

There are a few transactions that are marked null. Which shows that there are almost 698 rows of transactions which are marked nulls. Let's find out how many customers have these sets of unkown transactions.

In [157]:
len(unknownTransactionType_df.accountNumber.unique())

449

This shows that a customer who has atleast one unknown transaction, can also be traced again for another unknown transaction.

In [200]:
df.merchantCountryCode.value_counts()

US     778511
MEX      3143
CAN      2426
PR       1559
          724
Name: merchantCountryCode, dtype: int64

There are 754 null values in merchantCategoryCode

In [202]:
df.acqCountry.value_counts()

US     774709
         4562
MEX      3130
CAN      2424
PR       1538
Name: acqCountry, dtype: int64

There are 4562 null values in acqCountry

In [197]:
df.posEntryMode.value_counts()

05    315035
09    236481
02    195934
90     19576
80     15283
        4054
Name: posEntryMode, dtype: int64

There are 4054 null values in posEntryMode

In [204]:
df.posConditionCode.value_counts()

01    628787
08    149634
99      7533
         409
Name: posConditionCode, dtype: int64

In [205]:
df.merchantCategoryCode.value_counts()

online_retail           202156
fastfood                112138
entertainment            80098
food                     75490
online_gifts             66238
rideshare                51136
hotels                   34097
fuel                     23910
subscriptions            22901
auto                     21651
health                   19092
personal care            18964
airline                  15412
mobileapps               14990
online_subscriptions     11067
furniture                 7432
food_delivery             6000
gym                       2209
cable/phone               1382
Name: merchantCategoryCode, dtype: int64

No Nulls in merchantCategoryCode

In [206]:
df.currentExpDate.value_counts()

03/2029    5103
08/2024    5087
10/2023    5075
05/2027    5063
01/2021    5041
           ... 
02/2033    4276
02/2029    4260
02/2026    4234
12/2019    4028
08/2033    1874
Name: currentExpDate, Length: 165, dtype: int64

Didn't expect that there would be similar expiration dates

In [207]:
df.accountOpenDate.value_counts()

2014-06-21    33623
2014-09-30    13335
2014-05-22    11353
2012-10-09    10867
2014-10-02    10653
              ...  
2014-04-18        1
2011-10-16        1
2006-12-22        1
2006-05-12        1
2009-11-27        1
Name: accountOpenDate, Length: 1820, dtype: int64

No Null Values in accountOpenDate

In [209]:
df.dateOfLastAddressChange.value_counts()

2016-03-15    3819
2016-01-06    3740
2016-01-04    3558
2016-06-08    3355
2016-04-04    3194
              ... 
2009-11-27       1
2012-02-17       1
2009-05-14       1
2011-10-16       1
2011-05-09       1
Name: dateOfLastAddressChange, Length: 2184, dtype: int64

No Null Values

In [210]:
df.currentBalance.value_counts()

0.00       33678
8.46          26
5.67          19
29.42         19
59.68         18
           ...  
5075.51        1
3869.28        1
2350.72        1
3078.01        1
738.50         1
Name: currentBalance, Length: 487318, dtype: int64

 
No null values

In [212]:
df.cardPresent.value_counts()

False    433495
True     352868
Name: cardPresent, dtype: int64

There are no null values in cardPresent

In [213]:
df.expirationDateKeyInMatch.value_counts()

False    785320
True       1043
Name: expirationDateKeyInMatch, dtype: int64

There are no null values in expirationDateKeyInMatch

In [214]:
df.isFraud.value_counts()

False    773946
True      12417
Name: isFraud, dtype: int64

There are no null values in the target variable

#### [4.7]

**Assumption:** The following ones are interesting. After a slight look onto the dataframe, it appears that the entire record has null values.


**Method:** Using the same value counts, we find the null values for the following records.


In [215]:
print(df.echoBuffer.value_counts())
print(df.merchantCity.value_counts())
print(df.merchantState.value_counts())
print(df.merchantZip.value_counts())
print(df.posOnPremises.value_counts())
print(df.recurringAuthInd.value_counts())

    786363
Name: echoBuffer, dtype: int64
    786363
Name: merchantCity, dtype: int64
    786363
Name: merchantState, dtype: int64
    786363
Name: merchantZip, dtype: int64
    786363
Name: posOnPremises, dtype: int64
    786363
Name: recurringAuthInd, dtype: int64


Thus the entire record is of Null values. We can safely remove these from the consideration for using them in further process.

## [5] Conclusions:

Different sets of unique values were found using the unique feature and I could create a new dataframe for creditlimit and group by account numbers.

We are also able to identify the null values in every record of the dataset